block/src/qcow/mod.rs

```diff
@@ -2010,6 +2033,50 @@ mod tests {
         assert_eq!(read_header.backing_file_path, header.backing_file_path);
     }
 
+    /// Create a qcow2 file with itself as its backing file.
+    ///
+    /// Without configuration `max_nesting_depth`, this will cause infinite recursion when loading
+    /// the file.
+    fn new_self_referential_qcow(path: &Path) -> Result<()> {
+        let header = QcowHeader::create_for_size_and_path(3, 0x10_0000, path.to_str())?;
+        let mut disk_file = RawFile::new(
+            File::create(path).expect("Failed to create image file."),
+            false,
+        );
+        header.write_to(&mut disk_file)?;
+        Ok(())
+    }
+
+    #[test]
+    fn max_nesting_backing() {
+        let test_dir = TempDir::new_with_prefix("/tmp/ch").unwrap();
+        let img_path = test_dir.as_path().join("test.img");
+
+        new_self_referential_qcow(img_path.as_path()).unwrap();
+
+        let err = create_disk_file(
+            File::open(img_path.as_path()).expect("Failed to open qcow image file"),
+            false,
+            MAX_NESTING_DEPTH,
+        )
+        .expect_err("Opening qcow file with itself as backing file should fail.");
+
+        // The error is in a Box and cannot be compared directly. To make sure the error exists, we
+        // can check if it contains the type name after formatting.
```

Comment:
Without `Box`, types cannot be nested. The returned type is actually `Error` instead of `Box`, but `Box` is in `qcow::Error::BackingFileOpen`:

```rust
#[sorted]
#[derive(Debug, Error)]
pub enum Error {
    /* ...... */
    #[error("Backing file open error: {0}")]
    BackingFileOpen(Box<crate::Error>),
    /* ...... */
}
```

For such complex and nested errors, `matches!` does not work.